GLOSSARY.md

```diff
@@ -0,0 +1,128 @@
+# Glossary
+This is a glossary of terminology (possibly) used in the chalk crate.
+
+## Binary connective
+There are sixteen logical connectives on two boolean variables. The most
+interesting in this context are listed below. There is also a truth table given
+which encodes the possible results of the operations like this
+
+```
+f(false, false) f(false, true) f(true, false) f(true, true).
+```
+
+As a shorthand the resulting truth table is encoded with `true = 1` and `false =
+0`.
+
+```
+| Truth table | Operator symbol | Common name                      |
+|-------------|-----------------|----------------------------------|
+| 0001        | &&              | Conjunction; and                 |
+| 0111        | ||              | Disjunction; or                  |
+| 1001        | <=>             | Equivalence; if and only if; iff |
+| 1101        | =>              | Implication; if ... then         |
+```
+
+## Binder
+A binder is an expression that binds a variable to a certain expression.
+Examples for binders:
+
+- The universal quantifier `forall(a)` states that a certain condition holds for
+  all allowed values for `a`.
+- A function definition `f(x) = a * x` is a binder for the variable `x` whereas
+  `a` is a free variable.
+- A sum `\sum_n x_n` binds the index variable `n`.
+
+## Canonical
```

Comment:
So Canonical applies to an object which contains variables (e.g., a goal) and means that:
- we have replaced some variables with their value if any (i.e. we have applied our current "knowledge")
- we have rebound De Bruijn indices to lower ones

Chalk only deals with goals in canonical form.

For example, when dealing with a goal like: `forall<0,1> { 0: Foo && 1: Bar }`. Here we have two different indices because there are two different variables, and this goal is in canonical form.

In order to process it we will process each subgoal (namely `0: Foo` and `1: Bar`) separately, and so when we process `1: Bar`, this goal is not in canonical form because it could be written as `0: Bar` (lower De Bruijn indices). So we canonicalize this subgoal, process it, and then feed back the answer to the top level goal.